### ISSUE TITLE
How to ship assets (images, fonts) in a NPM package/library bundled by Webpack?

### ISSUE THREAD

**Issue by rdsedmundo** [webpack/webpack#7353]
**Note: I haven't created a question on StackOverflow, because in fact they already exist, two, and both don't have a good conclusion/solution for the problem, just a workaround, which is insufficient.**

**Those are the SO's questions:**

https://stackoverflow.com/questions/37171767/how-to-load-assets-correctly-from-webpack-generated-npm-package
https://stackoverflow.com/questions/36559850/include-assets-from-webpack-bundled-npm-package

I have the exact same problem as the 1) question. I want to be able to ship assets in an NPM package bundled by Webpack, but not necessarily I have control over the Webpack from the app who will use that library, i.e I can't just use the solution proposed on the 2) question.

The package is already well configured with `output.library`, `output.libraryTarget` set. The functionality of it works perfectly, the problem is only with their assets. I can see that they were outputted to the `dist` file inside of `node_modules/myLibrary` as expected using `file-loader`, but they aren't served correctly when the library require them, because they are in the `node_modules` of the library, not in the folder of the main app for instance.

The generated code of the library seems like (the part regarding the assets):

```javascript
/***/ "./src/assets/icons/add-tap-pixel.svg":
/*!********************************************!*\
  !*** ./src/assets/icons/add-tap-pixel.svg ***!
  \********************************************/
/*! no static exports found */
/***/ (function(module, exports, __webpack_require__) {

eval("module.exports = __webpack_require__.p + \"88112c78dd50d85634e6a85d3454bd9d.svg\";\n\n//# sourceURL=webpack://connections-component/./src/assets/icons/add-tap-pixel.svg?");
```

And when the library uses it, it tries to load the asset using the `publicPath` configuration as expected in the code via the `__webpack_require__.p` - the problem is just that the assets aren't there at all. They're in the `node_modules/myLibrary/dist`. And no, as I said, I can't just add a rule on the main app that uses the library to copy those files, because this library is supposed to be used in some other places which I not necessarily will have control over it build process.

I have knowledge that one solution would be to use `url-loader` for all assets, thus making them available directly via the JS code, but this is not what I'd like to do, as I want to have caching individually on each asset file.

Does anyone have a better solution for this? Thanks!

**Comment by rdsedmundo** [comment]
@sokra @jhnns @TheLarkInn @SpaceK33z 
Can any of you just confirm if there's a solution or not for this? Because if it's a problem that needs to be fixed, I'm going to remove webpack from the library package and just use a normal transpiler instead.

Thanks.

**Comment by ooflorent** [comment]
**Disclaimer:** The solution for this is really opinionated and there is no best answer since each use-case is different. Here are a few possibilities:

1. **Use `url-loader` without `limit` for assets or `raw-loader` for content.**
  This works well if you have _small_ files.

2. **Let the user decide how to include the asset.**
  Include the assets in your npm package and don't use any loader to bundle them. The end user has to configure its build system to ship the assets.

**Comment by ooflorent** [comment]
Solution 2 is valid. Here are few examples:

- https://github.com/quilljs/quill (18,185 ⭐️)
- https://github.com/gpbl/react-day-picker (2,146 ⭐️)

**Comment by yang-jandro** [comment]
@rdsedmundo were you able to get this working?

**Comment by rdsedmundo** [comment]
No. 2. didn't work for me as Webpack says that it doesn't know how to handle the file. I had to stick with `url-loader`.

**Comment by yang-jandro** [comment]
Ah okay, yeah after some research, ended up having to do the same thing.

**Comment by jekelija** [comment]
@ooflorent Looking at quill, it is using html-loader to bundle inline. I also had to stick with url-loader.

**Comment by rikbrowning** [comment]
I think the main issue is around defining the public path. If I create a library that has a component that uses an image it will construct a path based off the public path defined when the library is built. Ideally when consuming that library you want to update the public path to be whatever is now appropriate for that consumer. They can then include the image as necessary to get it output to the new public path. The issue is the library is not aware of this change and still looks at the old path.  Anyone aware of a good way to handle this situation?

**Comment by yang-jandro** [comment]
If it's worth anything, my library was for private use and was always going to be consumed and pre-bundled. With that said, I moved away from using webpack to actually bundle the library and instead just transpiled and copied assets over directly. That way, images will stay in relative paths, and the projects consuming are able to properly bundle using webpack (or any other bundler). Worked like a charm.

**Comment by rdsedmundo** [comment]
@ooflorent Can you please re-open the issue or explain the `2` solution you gave that doesn't worked for all of us? As I stated once you closed the issue, the solution you proposed is not valid as `webpack` will claim that it can't resolve the file, if the loader gets removed.

**Comment by ooflorent** [comment]
### tl;dr

- **Solution 1:** use `url-loader` so your consumers never know about the assets
- **Solution 2:** you don't care how your consumer includes your assets, that's their problem

---

Let's recap the use-case: **we want to distribute some assets along with our library**. Here are some restrictions for our solution:
- it must not require the library consumer to use webpack
- it must be easy for the library consumer to include the assets

### The problem

Using `file-loader` ties the library code to a public path which is hard to configure. Moreover, I would require the library consumer to use webpack or its runtime.

### Solution 1: don't ship the assets, _embed_ them

By using `url-loader` all matching assets will be included as URLs inside the final JS bundle. A simple configuration could be:

```js
{
  test: /\.(jpg|png|svg)$/,
  loader: "url-loader",
  options: {
    limit: Infinity // everything
  }
}
```

### Solution 2: externalize assets

Instead of bundling the assets inside the library, they are distributed as separate files of your package. For example:

```
my-module
├── dist
│   └── bundle.js
├── styles.css
└── package.json
```

Here, `styles.css` is not produced by webpack. It is not referenced from any file of your bundle. The library consumers must manage this file themselves! So they want to use `my-module`? Perfect! They must `require("my-module")` and load `my-module/styles.css` whatever is there preferred solution.

Personally I would use `css-loader`:

```js
import "my-module/styles.css"
import MyModule from "my-module"
```

This is not the library business to handle this.

--- 

> Can you please re-open the issue […]?

No I won't because your use-case is very localized and would be better handled on StackOverflow.

**Comment by rikbrowning** [comment]
Solution 1 is not ideal for large images.
Solution 2 does not work for images as if the module needs to know where the image is it has already been built with a public path. 
Ideally we would need to treat the image as external then when the consuming app or module reaches that require it could handle it via a loader. Then solution 1 or 2 could come into play.

**Comment by rdsedmundo** [comment]
7 different people either reacted or commented here looking for the same solution, but my issue is "very localized". You shall be correct. 👍

**Comment by sokra** [comment]
There is no perfect solution for this problem yet. There is some ongoing spec work to define a syntax for asset imports, but that is not finished yet.

My recommendation: don't bundle packages before pushing them to npm. Ship the raw or transpile-only source. Reference assets via require or import, most bundlers understand that. Maybe add some hints to readme how to configure bundlers to allow to use it.

You may offer an additional bundled version for `<script>` usage. You could allow setting publicPath via data tag on the script tag (`__webpack_public_path__ = document.currentScript.dataset.assetsPath`).

**Comment by elado** [comment]
Been contemplating how to do the same thing myself. Not sure I'd call this use case "localized" because currently, there's a hard limitation on including images/fonts in packages, without embedding them as base64.

`url-loader` is not optimal for bigger images, and `file-loader` cannot be used due to a different `publicPath` that the consumer app may have.

For example:

```jsx
import React from 'react';
import img from './images/1.png';

export const Comp = () => <div><img src={img} /></div>;
```

`file-loader` puts the file in the dist folder and hard codes `publicPath` to access it (e.g. `img = "/assets/1.<hash>.png'"`)

The consumer of this package cannot control this internal publicPath and/or easily get all the output assets.

With CSS files it's easy to instruct the consumer to load it in a `import`, but assets that use `file-loader` turn into a string path, which won't be compatible with the consuming app.

It seems to me like a new loader/enhanced file-loader, and plugin can solve it:

1. `file-loader` with a flag, that uses a dynamic `publicPath` instead of a string - and used in the shared package, will produce urls like `__globalPublicPath() + '1.<hash>.png'`
2. a plugin on the consumer side that a) injects `__globalPublicPath` with the current `publicPath`, and b) grabs the assets from the package and adds them into the dist folder.

This also needs to work with a nested packages that include assets.

This will also allow uploading these images to s3 with S3Plugin as a part of the consumer app.

**Comment by Halt001** [comment]
For those trying: **Solution 1:** don't ship the assets, embed them.

In the code sample below `Infinity` is not working for me as a limit, setting it to `true` works for infinite. See [the docs](https://github.com/webpack-contrib/url-loader#boolean).

```
{
  test: /\.(jpg|png|svg)$/,
  loader: "url-loader",
  options: {
    limit: Infinity // everything
  }
}
```

**Comment by davidjungermann** [comment]
Has any work been done in regards to this issue? 

Would love an easier way to bundle and ship assets. We are currently using a single repository for documentation in Storybook.js, based on Create-React-App and webpack to bundle and ship components as npm modules. If assets could be loaded using the same path, like @rikbrowning pointed out, it would make life a lot easier for us.

**Comment by michaelknoch** [comment]
is there any public example of approach 2) with the `publicPath` property @elado mentioned?

**Comment by AllanOricil** [comment]
I have this problem. I created a package that draws on canvas. The canvas has a background image, which the package bundled. I used file-loader to bundle the image inside the package. Then the package alone can figure out where the image is. Like, I created a index.html and loaded the index.js script in the html to test, and it worked. 
But when I use this package as a dependency in another project, for instance Nuxt JS, the background image can't be found. Nuxt is expecting the image to be in the "localhost:3000/assets/images/image.svg". Nuxt is buit using Webpack and it could not bring my image in the resulted bundle.  Why????

**Comment by thw0rted** [comment]
I haven't seen anybody here mention the `exports` field in `package.json`.  The [webpack docs](https://webpack.js.org/guides/package-exports/) reference the [node docs](https://nodejs.org/api/packages.html#packages_package_entry_points), and from the "Conditions" heading, it looks like you can use the `asset` conditional to expose e.g. fonts for stylesheets to find.  I'm just starting to mess with this now, so I'd be interested to hear from anybody with more experience.

**Comment by trusktr** [comment]
> Reference assets via require or import, most bundlers understand that

Some libraries have

```js
import * as image from './image.jpg'
```

in their distributed source, and others have


```js
import image from './image.jpg'
```

How does the appplication with its bundler (f.e. Webpack) handle the discrepancy?

**Comment by trusktr** [comment]
Is there a way to just write standards-based `await fetch('./path/to/some/asset.png')` and then have (end user's) Webpack replace the URL with the final file's path or base64 URL, instead of relying on not-yet-standard ESM for assets?

**Comment by alexander-akait** [comment]
You can use new ES syntax: `console.log(new URL('./file.ext', import.meta.url));`, for Node.js example:
```
import fs from 'fs';

const pathToFile = new URL('./package.json', import.meta.url);

console.log(fs.readFileSync(pathToFile));
```

**Comment by trusktr** [comment]
Oh neat. Works in Webpack?

**Comment by alexander-akait** [comment]
@trusktr in webpack@5, new built-in feature

**Comment by thw0rted** [comment]
I don't understand how `import.meta.url` helps with @trusktr 's question about making assets available.

Let's say I built a library with:

````js
// node_modules/checkmark/main.js
export function decorate(selector) {
  const img = document.createElement("img");
  const srcPath = new URL("./check.png", import.meta.url);
  img.src = srcPath.toString();
  document.querySelector(selector).appendChild(img); // it's probably fine
}
````

which my application consumes as

````js
import {decorate} from "checkmark";

decorate("#checkme");
````

Now, I bundle my application with Webpack, using a config that has rules for JS through `babel-loader` and PNG as an `asset`.

What path will the `<img>` tag that gets attached to the `#checkme` element have when I load the resulting bundle in a browser?  Will Webpack do anything with `node_modules/checkmark/check.png` by default (copy it, or inline it, maybe depending on its size), or does the consuming project have to put something like `copy-webpack-plugin` in the config?

(Note that his question was about using `fetch`, not `import` or `require`.  My example isn't `fetch`-based, but the point was to pass the asset path to something that isn't statically analyzed by webpack.)

**Comment by alexander-akait** [comment]
> What path will the <img> tag that gets attached to the #checkme element have when I load the resulting bundle in a browser? Will Webpack do anything with node_modules/checkmark/check.png by default (copy it, or inline it, maybe depending on its size), or does the consuming project have to put something like copy-webpack-plugin in the config?

No need something in configuration(s), webpack will resolve asset and convert this to code (and rename image based on configuration or not) to URL address (i.e. `publicPath` + `assetsFilename`). 

> (Note that his question was about using fetch, not import or require. My example isn't fetch-based, but the point was to pass the asset path to something that isn't statically analyzed by webpack.)

Feel free to using fetch in this case `fetch(new URL("./check.png", import.meta.url))`, but don't forget `fetch` is not supported by Node.js (but you can polyfill it).

**Comment by thw0rted** [comment]
How does this work under the hood?  Is it documented somewhere?  Is the `URL` constructor actually hooked by Webpack to do static analysis and add assets to the pipeline?  That is, if I just write `new URL("./somefile.json", import.meta.url)` anywhere in my library code, does that result in `somefile.json` being handled as [`type: "asset"`](https://webpack.js.org/guides/asset-modules/), the way it would for `require()` or `import`?  (It seems like this would be harder to hook than `require`/`import`, because I can also write `let metaUrl = import.meta.url; metaUrl += "/subpath/"; const someFile = new URL("./somefile.json", metaUrl);`, etc.)

Does this also mean that using a library that includes `new URL("./somefile.xyz", import.meta.url)` will cause bundler errors because the consuming project doesn't have a rule declared for the `.xyz` extension?  Or is there some kind of default / fallback behavior?

**Comment by alexander-akait** [comment]
> How does this work under the hood? Is it documented somewhere? Is the URL constructor actually hooked by Webpack to do static analysis and add assets to the pipeline?

https://webpack.js.org/guides/asset-modules/#url-assets, but yes we should improve this

> Is the URL constructor actually hooked by Webpack to do static analysis and add assets to the pipeline?
> That is, if I just write new URL("./somefile.json", import.meta.url) anywhere in my library code, does that result in somefile.json being handled as type: "asset", the way it would for require() or import?

yes, look at link

>  (It seems like this would be harder to hook than require/import, because I can also write `let metaUrl = import.meta.url; metaUrl += "/subpath/"; const someFile = new URL("./somefile.json", metaUrl);`, etc.)

Some edge cases doesn't work, because it is very new, but yes, we will improve it after feedback (feel free to open an issue with you example), in most cases you don't need to touch `import.meta.url`

> Does this also mean that using a library that includes new URL("./somefile.xyz", import.meta.url) will cause bundler errors because the consuming project doesn't have a rule declared for the .xyz extension? Or is there some kind of default / fallback behavior?

No, all `new URL('./file.ext', import.meta.url)` are loaded as URL, you don't need extra rules or configurations, all works out of box, but if you need compress or change content of assets you need `loader` here

**Comment by thw0rted** [comment]
This is really great, thanks.  I will play around with it for my own use-case when I get a chance.  It's exciting to have a way forward that will work natively cross-platform (once all the implementers get on board with the `import.meta` spec...)

**Comment by alexander-akait** [comment]
@thw0rted feel free to feedback :sta

**Comment by NilsBaumgartner1994** [comment]
> **Disclaimer:** The solution for this is really opinionated and there is no best answer since each use-case is different. Here are a few possibilities:
> 
> 1. **Use `url-loader` without `limit` for assets or `raw-loader` for content.**
>    This works well if you have _small_ files.
> 2. **Let the user decide how to include the asset.**
>    Include the assets in your npm package and don't use any loader to bundle them. The end user has to configure its build system to ship the assets.

Thanks the url-loader worked like a charm

**Comment by karlhorky** [comment]
Since I didn't find this anywhere, here's a demo I made of the new `new URL('./file.png', import.meta.url)` syntax with ES Modules without a bundler:

https://codesandbox.io/s/import-image-paths-in-js-without-bundler-cbqpk

**Comment by thw0rted** [comment]
Hey @alexander-akait , I linked to this from https://github.com/PolymerLabs/playground-elements/issues/131 -- it looks like they're having problems with this construct.  It looks like they're creating an effectively unbounded context, by passing an arbitrary string variable to the first argument of `new URL(...)`.  What's supposed to happen there?  I actually have a similar problem with another library I'm using, where they wrap a library-relative resource reference in a helper function and thus prevent static analysis of the (in this case, `require`) call.  Seems like a hard problem to address.

**Comment by gretzkiy** [comment]
@alexander-akait hi! I am trying to use the `new URL` mechanism to bundle library with assets (and not inlining them), but I can't make it work. The problem is that webpack doesn't emit the asset into the bundle of the project that uses that library.

I created a [repo](https://github.com/gretzkiy/webpack-url-assets-test) with very simple example of the library with image as asset and the project that consumes that ibrary. The library uses the image with `new URL` syntax and the image is bundled in library's `dist` directory. But when I use the library in project the image is not bundled in project's `dist` and so the image loading is broken because it doesn't exist in the final build. 

So my question is: do I miss something or it shouldn't work at all in that way?

**Comment by alexander-akait** [comment]
@gretzkiy we have an issue about it, but I recommend do not bundle library, `import` and `new URL(...)` already works in modern browsers, developers which want to support old browser will be bundle and webpack (or other bundlers) will handle `new URL` in your library

**Comment by gretzkiy** [comment]
Sounds reasonable, thanks. However could you please provide the link to the issue you mentioned?

**Comment by alexander-akait** [comment]
@gretzkiy https://github.com/webpack/webpack/issues/12719, here problem about `new Worker(new URL('./worker.js', import.meta.url));`, but the problem is the same, workers doesn't copied

**Comment by gretzkiy** [comment]
@alexander-akait many thanks

**Comment by mixtur** [comment]
In esbuild there is a `copy` loader. That solves the problem in the following way: they copy the file as file-loader or asset/resource would, but then don't create a URL-ish string in the code. Instead they rewrite the import so it would point to the new relative location of the file in the output directory.
Is there a way to achieve this behavior in webpack?
